hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -0,0 +1,208 @@
+import javax.annotation.Nonnull;
+import java.security.AccessControlException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Client-side implementation of SQL service.
+ */
+public class SqlClientService implements SqlService {
+
+    private final HazelcastClientInstanceImpl client;
+
+    public SqlClientService(HazelcastClientInstanceImpl client) {
+        this.client = client;
+    }
+
+    @Nonnull
+    @Override
+    public SqlResult query(@Nonnull SqlQuery query) {
+        Connection connection = client.getConnectionManager().getRandomConnection(true);
+
+        if (connection == null) {
+            throw rethrow(QueryException.error(
+                SqlErrorCode.CONNECTION_PROBLEM,
+                "Client must be connected to at least one data member to execute SQL queries"
+            ));
+        }
+
+        try {
+            List<Object> params = query.getParameters();
+
+            List<Data> params0 = new ArrayList<>(params.size());
+
+            for (Object param : params) {
+                params0.add(serializeParameter(param));
+            }
+
+            ClientMessage requestMessage = SqlExecuteCodec.encodeRequest(
+                query.getSql(),
+                params0,
+                query.getTimeoutMillis(),
+                query.getCursorBufferSize()
+            );
+
+            ClientMessage responseMessage = invoke(requestMessage, connection);
+
+            SqlExecuteCodec.ResponseParameters responseParameters = SqlExecuteCodec.decodeResponse(responseMessage);
+
+            SqlExecuteResponse response = new SqlExecuteResponse(
+                    responseParameters.queryId,
+                    responseParameters.rowMetadata,
+                    responseParameters.rowPage,
+                    responseParameters.error
+            );
+
+            handleResponseError(response.getError());
+
+            return new SqlClientResult(
+                this,
+                connection,
+                response.getQueryId(),
+                response.getRowMetadata(),
+                response.getPage(),
+                query.getCursorBufferSize()
+            );
+        } catch (Exception e) {
+            throw rethrow(e, connection);
+        }
+    }
+
+    /**
+     * Fetch the next page of the given query.
+     *
+     * @param connection Connection.
+     * @param queryId Query ID.
+     * @return Pair: fetched rows + last page flag.
+     */
+    public SqlPage fetch(Connection connection, QueryId queryId, int cursorBufferSize) {
+        try {
+            ClientMessage requestMessage = SqlFetchCodec.encodeRequest(queryId, cursorBufferSize);
+            ClientMessage responseMessage = invoke(requestMessage, connection);
+            SqlFetchCodec.ResponseParameters responseParameters = SqlFetchCodec.decodeResponse(responseMessage);
+            SqlFetchResponse response = new SqlFetchResponse(responseParameters.rowPage, responseParameters.error);
```

Comment:
There is no need for `SqlFetchResponse ` object anymore here.

Reply:
Fixed.